### ISSUE TITLE
Throw exception on mockStatic if the class wasn't prepared for test (@PrepareForTest(ClassWithStatic.class)).

### ISSUE THREAD

**Issue by astafev** [jayway/powermock#604]
In this scenario:
```
mockStatic(ClassWithStatic.class);
when(ClassWithStatic.someString()).thenReturn("asdf");
```
If ClassWithStatic wasn't prepared for test (I mean annotation @PrepareForTest(ClassWithStatic.class)), exception will be thrown only on 2nd line.
It's a but confusing, would be better if the exception would be thrown on mockStatic invocation.

**Comment by johanhaleby** [comment]
Good idea, please provide a pull request.

**Comment by astafev** [comment]
Hello, I added the check, but not sure how to act in some circumstances, could you take a look at the test samples.powermockito.junit4.partialmocking.PartialMockingRetainsStateTest#spyingOnAClassRetainsState? Should exception be thrown in that case?

One more thought: this check may break backward compatibility in some cases like the aforementioned. Maybe it's worth to make it configurable, like MockitoConfiguration in mockito?

**Comment by thekingnothing** [comment]
@astafev, from my point of view, in case Spy exception shouldn't be thrown, because it doesn't break any logic if a class isn't prepared for test. 

With `spy` there is another case, then I think exception should be thrown, because otherwise test will be passed, then it really isn't passed. 

Example: 

	@Test
	public void spyingOnAClassRetainsStateVerify() throws Exception {
		PowerMockito.spy(MockWithStaticStateDemo.class);
		PowerMockito.verifyStatic();
		MockWithStaticStateDemo.getState();
	}


This test is passed without `@PrepareForTest(MockWithStaticStateDemo.class)`. 

In bottom line, my thoughts, exception should be thrown in follow case: 

- Use precondition for static/private without preparing class, like `when`, 'given' and so on. 
- As is pointed in the issue, in case call `mockStatic` without preparing class 
- call `verifyStatic` without preparing class

**Comment by astafev** [comment]
Hi, I don't see big difference between spying and mocking, because spy won't work as well. In your example tests would fail on next verification. Like this:
```
@Test
public void spyingOnAClassRetainsStateVerify() throws Exception {
    PowerMockito.spy(MockWithStaticStateDemo.class);
    assertEquals(5, MockWithStaticStateDemo.getState());
    PowerMockito.verifyStatic();
    MockWithStaticStateDemo.getState(); //no verification made here in fact, real method is invoked
    validateMockitoUsage(); // exception is thrown here
}
```
> Use precondition for static/private without preparing class, like when, 'given' and so on. 

In this case org.mockito.exceptions.misusing.MissingMethodInvocationException will be thrown in current implementation. (try to execute ```when(MockWithStaticStateDemo.getState()).thenReturn(3);``` in the test)

> call verifyStatic without preparing class

I don't see a way to make this check, because in that case no powermock api is executed.

**Comment by thekingnothing** [comment]
@astafev, 

I've thought again and I agree with you, that if a user tries to create spy for class and it means that it will be used for mocking or verifying static methods, then we should throw an exception and inform that `@PrepareForTest` should be used. 

> In this case org.mockito.exceptions.misusing.MissingMethodInvocationException will be thrown in current implementation. (try to execute when(MockWithStaticStateDemo.getState()).thenReturn(3); in the test)

Yes, exception is thrown in this case, but the exception does not inform a user about the root clause and it could confuse the user. If we can, better  approach will be catch this exception, check if it's static call and `mockStatic` isn't called then throw another exception and describe to user what the user should do to fix it.

**Comment by astafev** [comment]
@thekingnothing,
Sadly, I don't see a way to make the exception more informative. This problem is not only about static methods, it's about the whole mocking mechanism and the check is made inside mockito.

**Comment by thekingnothing** [comment]
@astafev, 

I mean, we have several cases then a user (developer) could incorrect use not fully prepared mock object and where it could be detected, base on pre-condition. 

1. Use `PowerMockito.mockStatic(MockWithStaticStateDemo.class)` without `@PrepareForTest` 
2. Use `PowerMockito.spy(MockWithStaticStateDemo.class)` without `@PrepareForTest`
3. Use `when(MockWithStaticStateDemo.getState())` without `@PrepareForTest` or `PowerMockito.mockStatic(MockWithStaticStateDemo.class)`
4. Use `PowerMockito.verifyStatic(MockWithStaticStateDemo.class)` without `@PrepareForTest` or `PowerMockito.mockStatic(MockWithStaticStateDemo.class)` - for mocks. 

I age that it should be implemented in different piece of code, but in all these cases behaviour should be unified: throw exception with full information what the user has done wrong and how it could be fixed.

**Comment by thekingnothing** [comment]
It could be separated into several issue and I can create a label/branch for this feature, but I strong believe that all these cases should be covered together and release in one release. 

This is a excellent idea. We exactly need to add this feature. I really get tired to explain same mistakes for all newcomers.

**Comment by thekingnothing** [comment]
@astafev, I've create a new branch for this feature 
https://github.com/jayway/powermock/tree/powermock-629

Could you create a pull request with your change?

**Comment by thekingnothing** [comment]
Fix in #629
